framework/src/main/java/org/checkerframework/common/wholeprograminference/scenelib/ASceneWrapper.java

```diff
@@ -0,0 +1,251 @@
+        Set<String> annosToIgnoreForLocation = annosToIgnore.get(Pair.of(annosToIgnoreKey, loc));
+        if (annosToIgnoreForLocation != null) {
+            Set<Annotation> annosToRemove = new HashSet<>();
+            for (Annotation anno : typeElt.tlAnnotationsHere) {
+                if (annosToIgnoreForLocation.contains(anno.def().toString())) {
+                    annosToRemove.add(anno);
+                }
+            }
+            typeElt.tlAnnotationsHere.removeAll(annosToRemove);
+        }
+
+        // Recursively remove ignored annotations from inner types
+        for (ATypeElement innerType : typeElt.innerTypes.values()) {
+            removeIgnoredAnnosFromATypeElement(innerType, loc, annosToIgnore);
+        }
+    }
+
+    /**
+     * Write the scene wrapped by this object to a .jaif file on the given path.
+     *
+     * @param jaifPath the path to the jaif file to be written
+     * @param annosToIgnore which annotations should be ignored in which contexts
+     */
+    public void writeToJaif(
+            String jaifPath, Map<Pair<String, TypeUseLocation>, Set<String>> annosToIgnore) {
+        removeIgnoredAnnosFromScene(theScene, annosToIgnore);
+        theScene.prune();
+        new File(jaifPath).delete();
+        if (!theScene.isEmpty()) {
+            // Only write non-empty scenes into .jaif files.
+            try {
+                IndexFileWriter.write(theScene, new FileWriter(jaifPath));
+            } catch (IOException e) {
+                throw new UserError("Problem while writing %s: %s", jaifPath, e.getMessage());
+            } catch (DefException e) {
+                throw new BugInCF(e);
+            }
+        }
+    }
+
+    /**
+     * Write the scene represented by this object to a stub file.
+     *
+     * @param jaifPath a path that ends in ".jaif". The stub file will be created on the same path,
+     *     but the extension will be replaced with ".astub"
+     * @param annosToIgnore which annotations to ignore in which contexts
+     */
+    public void writeToStub(
+            String jaifPath, Map<Pair<String, TypeUseLocation>, Set<String>> annosToIgnore) {
+        removeIgnoredAnnosFromScene(theScene, annosToIgnore);
+        theScene.prune();
+        String stubPath = jaifPath.replace(".jaif", ".astub");
+        new File(stubPath).delete();
+        if (!theScene.isEmpty()) {
+            // Only write non-empty scenes into .astub files.
+            try {
+                SceneToStubWriter.write(this, new FileWriter(stubPath));
+            } catch (IOException e) {
+                throw new UserError("Problem while writing %s: %s", stubPath, e.getMessage());
+            }
+        }
+    }
+
+    /**
+     * Interact with scenelib to add a class to the scene.
+     *
+     * <p>Results are interned.
+     *
+     * <p>updateClassMetadata has to be called on both paths (cache hit and cache miss) because the
```

Comment:
This belongs as a code comment.  Remarks about paths in the code are not part of the specification.  This can also be written more concisely.